docs: rewrite readme for @lavamoat/node (#1484)
LavaMoat/lavamoat: packages/node/README.md

# @lavamoat/node

> > Warning: You are looking at an early version of this packages. Breaking changes and bugs are to be expected.

Reviewer: ```suggestion
> [!WARNING]
>
> This software is _alpha_ quality. Expect bugs and breaking changes!
```

Reviewer: Omit L3 and L4 if you care about how it looks on npmjs.com

Author: yup, would rather keep npm looking nicer than github

Author: ```suggestion
> Warning: This software is _alpha_ quality. Expect bugs and breaking changes!
```

> **`@lavamoat/node` is a [Hardened JavaScript][] runtime for Node.js v18.0.0+ that provides per-package security policy enfocement**

Reviewer: ```suggestion
**`@lavamoat/node` is a [Hardened JavaScript][] runtime** for Node.js v18.0.0+ that provides _per-package_ security policy enforcement.
```

`@lavamoat/node`:

- Uses [`lockdown`][lockdown] to prevent tampering with the execution environment:
  - User code cannot meddle with JavaScript's global objects or shared intrinsics (i.e. "the
    stuff in the prototype chain")
  - Neutralizes most prototype pollution attacks
- Isolates dependencies _within the same process_:
  - By default, packages do not share references to any global objects
  - Access to resources (global objects, other packages, Node.js builtins,
    native modules) is controlled by user-defined policy

> - Provides tools for generating and maintaining the policy without the need to manually create all of it.

Reviewer: ```suggestion
- Provides tooling to _generate_ and _maintain_ the policy
```

- I don't think we need the last bit because:
  - at this point, the new reader does not know what a policy really is
  - nor do they understand what's involved in maintaining a policy 
  - and why a policy would or would not need to be created manually
- in a list, we can omit periods
- consistent tense

Author: thanks for the corrections. 
I still want to sell this as an upgrade from having to craft the whole policy manually, because I'm on a mission to leave hints of "you're responsible for the content of the policy file" everywhere.

LavaMoat can protect your Node.js application, but it can _also_ protect your development environment (e.g., build scripts) — you can run your tools with [the `lavamoat` CLI](#usage).

## Supply Chain Security

> Over time, your application's dependencies will get updated. It's easy to update to a version of a package that introduces malicious code or a new dependency on a malicious package.

Reviewer: ```suggestion
Over time, your application's dependencies will need updating. How can you ensure a new version of a package—or newly-added dependencies—aren't doing something malicious?
```

Asking a question like this is a pretty good hook into the section, imo

`@lavamoat/node` provides a runtime that should work just as before, but with protection against most supply chain attacks.

> Powerful APIs and Node builtins are available globally or through imports without restriction despite only very few dependencies of your application actually need to use them. LavaMoat can stop packages trying to abuse these resources without blocking their intended use.

Reviewer: ```suggestion
Nodes.js provides powerful global APIs and builtins without restriction—but most of your app's dependencies _don't need them_. LavaMoat can stop packages from abusing these resources _without_ blocking legitimate use.
```

> Each package can only access resources that are explicitly allowed by the policy. If a resource is not allowed for a package, the environment will behave as if that resource doesn't even exist. This is a powerful way to prevent malicious packages from doing harm.

Reviewer: ```suggestion
Each package can only access resources that are explicitly allowed by the policy. _If a package is disallowed access to a resource, the environment will behave as if that resource does not exist._ This is a powerful way to prevent malicious packages from doing harm.
```

thought this was worth emphasizing

Author: "resource is not allowed for a package" sounded more natural to me than "package is disallowed access to a resource" but I''m not the native speaker here :sweat_smile:

Reviewer: hmm, "if policy disallows access to a resource" is what I meant to write, I think.  active voice is less painful to read

Author: Then pls change that and merge if you don't mind. This stuff is ready to go.

## Runtime Security

> If a package obfuscates its intentions to the degree that LavaMoat's own policy generation didn't detect what it's using, the existing policy will prevent naughty behavior because LavaMoat doesn't rely on reading the code for the protections it provides. It makes sure resources not allowed by the policy for the package simply don't exist in the package's execution scope. (thanks to [Hardened JavaScript][] Compartments)

Reviewer: ```suggestion
If a package obfuscates its intentions to the degree that LavaMoat's own policy generation cannot detect what resources it needs, the generated policy will _still_ prevent naughty behavior. This is because LavaMoat operates using the [principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege). Resources **not explicitly allowed by the policy do not exist** in the package's execution scope—thanks to the `Compartment` provided by [Hardened JavaScript][].
```

I think it may not have been obvious that a policy is of the deny-by-default nature.  feel free to add URL to list at bottom of doc instead of inline

Author: very well put

> The only time LavaMoat reads the code is when generating a policy. Policy generation is there to provide a starting point for customization and it allows every resource it detects. You can use policy-overrides to restrict resources that are allowed for a package. And you need to review the policy you generate. (More on policy review in the [LavaMoat docs](https://lavamoat.github.io/guides/policy-diff/))

Reviewer: ```suggestion
LavaMoat parses code _only_ during policy generation—never at runtime. Policy generation provides a _starting point_ for customization by creating a policy which allows _every resource_ it detects. While reviewing the policy, you can choose to create **policy overrides** to restrict or grant access to resources.  

> [!TIP]
>
> Read more about reviewing policy files in the [LavaMoat docs](https://lavamoat.github.io/guides/policy-diff/).
```

using italics for emphasis and bold for terminology here. maybe it should be applied elsewhere

Reviewer: also pls add url at bottom of doc
